### ISSUE TITLE
Container hangs during normal use

### ISSUE THREAD

**Issue by tommflynn** [moby/moby#38005]
**Description**

Container hangs and is totally uncontactable. This occurs during "normal use", while I am running a test suite against an image of the product I work on. There is not one specific point or event which causes the hang (I've checked using trace) but during a 2-minute long test suite we hit this problem about 90% of times. 

**Describe the results you received:**
Docker container hangs. None of `docker kill [container]`, `docker exec [container] bash`, `docker stop [container]`, `docker logs [container]` have any effect nor return. `docker ps` shows the container, with correct created time, and status = up 'time since creation'. Restarting the VM running docker _does_ remove the container.

**Describe the results you expected:**
We would like the container not to hang, and we would like the container still to be possible to kill if it must hang. 

**Additional information you deem important (e.g. issue happens only occasionally):**
I have tried using sigkill to kill the container, and the container daemon using SIGHUP, -KILL and -USR1, none have any effect. 

**Output of `docker version`:**

```
Client:
 Version:      18.03.1-ce
 API version:  1.37
 Go version:   go1.9.5
 Git commit:   9ee9f40
 Built:        Thu Apr 26 07:20:16 2018
 OS/Arch:      linux/amd64
 Experimental: false
 Orchestrator: swarm

Server:
 Engine:
  Version:      18.03.1-ce
  API version:  1.37 (minimum version 1.12)
  Go version:   go1.9.5
  Git commit:   9ee9f40
  Built:        Thu Apr 26 07:23:58 2018
  OS/Arch:      linux/amd64
  Experimental: false

```

**Output of `docker info`:**

```
Containers: 11
 Running: 1
 Paused: 0
 Stopped: 10
Images: 5
Server Version: 18.03.1-ce
Storage Driver: overlay2
 Backing Filesystem: xfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 773c489c9c1b21a6d78b5c538cd395416ec50f88
runc version: 4fc53a81fb7c994640722ac585fa9ca548971871
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-862.3.3.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 3.701GiB
Name: pc5470-new-dev-env.ad.datcon.co.uk
ID: GYBZ:K5RS:MIF4:IZD4:PCK6:OABY:T2GV:FSQX:DRHZ:6OIV:QCKM:HNBN
Docker Root Dir: /home/.docker_data
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 pesdocker:5000
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Virtualbox Version 5.2.12 running CentOS Linux release 7.5.1804 (Core)

**Comment by kongbo1987** [comment]
I meet the same problem

**Comment by youssefNM** [comment]
same problem here

**Comment by cpuguy83** [comment]
Can you grab the goroutine stack trace from these stuck daemons?

You can get this with:
```
 curl --unix-socket /var/run/docker.sock http://./debug/pprof/goroutine?debug=2
```

**Comment by tdmb** [comment]
I'm a colleague of @tommflynn.
Here's the stack trace. Thanks for your help!
[stack_trace.log](https://github.com/moby/moby/files/2526384/stack_trace.log)

**Comment by cpuguy83** [comment]
Thanks, looks like we're stuck waiting on containerd to send a response to a kill request.
I noticed you are on 18.03 still, there are some known issues with the containerd shipped in that version.
Can you upgrade to 18.06?

**Comment by tdmb** [comment]
I'm afraid that I _was_ running 18.06 (which was later than @tommflynn - I hadn't spotted that he wasn't on the latest version). Since my setup seems to differ from his, I've added my `docker version` and `docker info` below.

**docker version**
```
Client:
 Version:           18.06.1-ce
 API version:       1.38
 Go version:        go1.10.3
 Git commit:        e68fc7a
 Built:             Tue Aug 21 17:23:03 2018
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.06.1-ce
  API version:      1.38 (minimum version 1.12)
  Go version:       go1.10.3
  Git commit:       e68fc7a
  Built:            Tue Aug 21 17:25:29 2018
  OS/Arch:          linux/amd64
  Experimental:     false
```

**docker info**
```
Containers: 5
 Running: 1
 Paused: 0
 Stopped: 4
Images: 6
Server Version: 18.06.1-ce
Storage Driver: overlay2
 Backing Filesystem: xfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 468a545b9edcd5932818eb9de8e72413e616e86e
runc version: 69663f0bd4b60df09991c08812a60108003fa340
init version: fec3683
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-862.11.6.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 10.34GiB
Name: <removed>
ID: DCLD:WJTC:I7RR:QMZ3:DGOQ:MCZ6:OG5N:JB7C:FHT2:3INV:HBKE:5X34
Docker Root Dir: /home/.docker_data
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 pesdocker:5000
 127.0.0.0/8
Live Restore Enabled: false

WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
```

**Comment by mo-mughrabi** [comment]
Same problem here

**Comment by mohamedrez** [comment]
Same problem here

**Comment by lewisdrummond1** [comment]
I have this problem too. Was it introduced with 18.03?

**Comment by abumalick** [comment]
We met the same problem

**Comment by cpuguy83** [comment]
Ok... Please don't come and say "we have the same problem".
Can't fix the problem unless you provide the details requested, which may not actually be the same problem just the same symptoms.

**Comment by tdmb** [comment]
@cpuguy83 - was my stack trace and info helpful? I'm happy to gather extra diags if you like - it repros really easily for us. Note that the [stack trace|https://github.com/moby/moby/files/2526384/stack_trace.log] was taken at the latest version 18.06

**Comment by lewisdrummond1** [comment]
```
docker version
Client:
 Version:           18.06.1-ce
 API version:       1.38
 Go version:        go1.10.3
 Git commit:        e68fc7a215d7133c34aa18e3b72b4a21fd0c6136
 Built:             Tue Sep 18 17:46:50 2018
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.06.1-ce
  API version:      1.38 (minimum version 1.12)
  Go version:       go1.10.3
  Git commit:       e68fc7a/18.06.1-ce
  Built:            Tue Sep 18 17:48:16 2018
  OS/Arch:          linux/amd64
  Experimental:     false
```
[stack_trace.log](https://github.com/moby/moby/files/2530915/stack_trace.log)

**Comment by cpuguy83** [comment]
@tdmb Immensely, thank you. By chance are you able to get the containerd stack trace as well? `docker-containerd-ctr pprof --debug-socket=/run/docker/containerd/docker-containerd-debug.sock goroutines`

**Comment by cpuguy83** [comment]
@lewisdrummond1 Can you describe what's blocked? I don't see anything in particular stuck in the stack trace.

**Comment by tdmb** [comment]
I've reproed again - here are the docker and containerd stack traces.
[docker_stack_repro1.log](https://github.com/moby/moby/files/2534640/docker_stack_repro1.log)
[containerd_stack_repro1.log](https://github.com/moby/moby/files/2534641/containerd_stack_repro1.log)

Again, please let me know what further diags I could provide, or debugging steps I could perform.
We're keen to get this resolved, or find a way to work around it.

**Comment by cpuguy83** [comment]
@tdmb Thanks
This one looks a bit different. Specifically it's stuck killing the container process, even containerd seems to be waiting for the containerd-shim (container process monitor) to exit or send something back to containerd.

I'm not sure at the moment if there is a better way to deal get this (sorry!)... but if you send SIGUSR1 to the shim process (which should be the parent process of the stuck container) this will generate a stack dump for the shim process in the docker daemon logs.

**Comment by cpuguy83** [comment]
@tdmb Also by chance are you using the `docker pause` or cgroup freezer functionality?

**Comment by tdmb** [comment]
That's not doing anything. There's no output in `/var/log/messages` when I send that signal to the `docker-containerd-shim` process (and I've looked in the [other possible locations] (https://docs.docker.com/config/daemon/#read-the-logs))
There are some docker logs in that file - so I've uploaded the recent logs from it
[messages.log](https://github.com/moby/moby/files/2535631/messages.log).
I sent the signal at "Oct 31 17:21:03", so I thought originally that the "ignoring event" message might be caused by this, but it's not from the shim, and nothing got logged when I sent the signal again (multiple times).

**Comment by cpuguy83** [comment]
@tdmb I'm wondering if this might be related to https://github.com/containerd/containerd/pull/2597
Would you be willing to drop in a custom containerd (and supporting) binary?

**Comment by tdmb** [comment]
Sure, if you give me instructions, and it's something I can revert easily afterwards.
Is this a debug build to get extra diags for you, or a build with a patch to fix containerd/containerd#2597?

**Comment by tdmb** [comment]
Unfortunately, that didn't fix the problem. My test suite did run further than it usually would (though that is likely a conincidence - I only ran it once), but then it ran into the same symptoms.

Here's the docker and containerd stacks again:
[docker_stack_containerd_2597.log](https://github.com/moby/moby/files/2554087/docker_stack_containerd_2597.log)
[containerd_stack_containerd_2597.log](https://github.com/moby/moby/files/2554086/containerd_stack_containerd_2597.log)

**Comment by cpuguy83** [comment]
Sadly same thing, stuck on kill.
Still no luck sending `SIGUSR1` to the stuck `containerd-shim` process? (note that each container gets it's own `containerd-shim`).

**Comment by cpuguy83** [comment]
@tdmb Do you typically run lots of execs (e.g. `docker exec`)?

**Comment by tdmb** [comment]
Still no luck sending `SIGUSR1` to the stuck `containerd-shim` process, I'm afraid. (I'm only running the stuck container at the moment, so it's definitely the right one),

We run a few number of `docker exec` commands to copy files around at the start, but not at the point that we run into this problem. We have a bunch of volumes set up so we can access directories outside the container, if that's interesting.

We've narrowed down the problem a bit - it seems to happen in the following scenario:
We have two processes each running in separate containers (A & B) with a TCP connection between them. We SSH on to container A and execute a shell script which kills and restarts the process. When container B notices that the TCP connection has gone, it calls `ipaddr add` to take control of A's IP address. As A's process restarts and re-establishes the TCP connection, it calls `ipaddr del`  to lose control of its old IP address. At some point shortly after A's process has restarted (perhaps before it calls `ipaddr del` or even re-establishes the TCP connection - we're not sure), we run into this issue.

It doesn't happen every time - but over ~10 of these restarts we tend to run into it.

**Comment by tdmb** [comment]
@cpuguy83 - do you have any suggestions of how we can help you debug this issue, or diags we can gather?

**Comment by tdmb** [comment]
We've managed to get to the root cause - we have a zombie process, which is preventing docker from killing the container. I believe we might be able to use the `--init` option to avoid that, but this is probably just how docker works (if not, I'm sure it's covered by another issue!).

**Comment by cpuguy83** [comment]
@tdmb Thanks for the extra details. Unfortunately I cannot reproduce this issue by using zombie processes.
Did the issue go away when you dealt with the zombies?

There are a number of fixes going into containerd around deadlocks, but it's difficult to tell what's going on here.

**Comment by tdmb** [comment]
Yes, the container can be killed by docker once the parent of the zombie process was sent a SIGKILL (which terminated both it and its child zombie). Unless you want to do any further investigation, we're happy to close this issue. @tommflynn - can you do that, please? @cpuguy83 - thank you very much for your help here.

**Comment by tommflynn** [comment]
Yep!

**Comment by cpuguy83** [comment]
Still curious how it's reproduced because docker should not lock up just because of a zombie process.
